aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -264,74 +327,41 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
   auto offset2bag_data = offset2bag.data<int64_t>();
   int64_t numel = indices.numel();
 
-  std::vector<int64_t> counts(num_weights);
-  for (int i = 0; i < numel; i++) {
-    counts[indices_data[i]] = 0;
-  }
-  for (int i = 0; i < numel; i++) {
-    counts[indices_data[i]]++;
-  }
-
-  auto index_grad_weight =
-      at::zeros({num_weights, grad.size(1)}, grad.options()).contiguous();
-
-  std::vector<int64_t> counts_uniq;
-  counts_uniq.reserve(num_weights);
-  int64_t o = 0;
-  for (int64_t i = 0; i < numel; i += counts[indices_data[i]]) {
-    counts_uniq.push_back(counts[indices_data[i]]);
-    if (o > 0) {
-      counts_uniq[o] += counts_uniq[o - 1];
-    }
-    o++;
-  }
-
-  if (mode == MODE_MEAN || mode == MODE_SUM) {
-    #pragma omp parallel for if (numel > 1000)
-      for (int64_t i = 0; i < (int64_t)counts_uniq.size(); i++) {
-        int64_t start = i == 0 ? 0 : counts_uniq[i - 1];
-        int64_t index = indices_data[start];
-        for (int64_t j = start; j < counts_uniq[i]; j++) {
-          int64_t source = offset2bag_data[j];
-          double scale = 1.0;
-          if (scale_grad_by_freq) {
-            scale /= counts[indices_data[i]];
-          }
-          if (mode == 1) { // MODE_MEAN
-            if (offsets_.size(0) == 1) {
-              auto bag_size = indices.size(0);
-              scale /= bag_size;
+  auto counts = compute_counts(num_weights, indices_data, numel);
+  auto next_unique_index_idx =
+      compute_counts_uniq(num_weights, indices_data, numel, counts);
+
+  #pragma omp parallel for if (numel > 1000)
```

Comment:
It's preferable to use parallel.h, if you want to fix more issues with this code.

Reply:
I don't know what parallel.h is, but this was here before and this pull request is a refactor of the code

Reply:
Yes, that was added after the last time I touched this file: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Parallel.h

Reply:
Going to bootcamp this...